### DISCUSSION TITLE
Populate doesn't work inside explicit transactions

### DISCUSSION THREAD

**Discussion by zdb999** [mikro-orm/mikro-orm#3218] [Q&A] [answered]
**Describe the bug**
When inside an explicit transaction, `em.populate()` throws an error. Additionally, the same error occurs when `populate` options are specified within a `find()` function or a `findOne()` function. It seems that populate is trying to manage its own transaction, which doesn't work, at least with the Postgres driver.

**Stack trace**
```
DriverException: Transaction query already complete, run with DEBUG=knex:tx for more info
    at PostgreSqlExceptionConverter.convertException (/home/my-repo/node_modules/@mikro-orm/core/platforms/ExceptionConverter.js:8:16)
    at PostgreSqlExceptionConverter.convertException (/home/my-repo/node_modules/@mikro-orm/postgresql/PostgreSqlExceptionConverter.js:42:22)
    at PostgreSqlDriver.convertException (/home/my-repo/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:182:54)
    at /home/my-repo/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:186:24
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at PostgreSqlDriver.find (/home/my-repo/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:46:24)
    at SqlEntityManager.find (/home/my-repo/node_modules/@mikro-orm/core/EntityManager.js:105:25)
    at EntityLoader.populateMany (/home/my-repo/node_modules/@mikro-orm/core/entity/EntityLoader.js:146:22)
    at EntityLoader.populateField (/home/my-repo/node_modules/@mikro-orm/core/entity/EntityLoader.js:217:25)
    at EntityLoader.populate (/home/my-repo/node_modules/@mikro-orm/core/entity/EntityLoader.js:37:13)
```

**To Reproduce**
1. Create an entity to with one-to-many or many-to-many relationships.
2. Start an explicit translation with `em.begin()`. Create a `try` block with appropriate commits and rollbacks. (I essentially recreated my own `em.transactional()` so I could wrap the transaction with some auth and auditing bookkeeping.)
3. Try to `findOne` of those entities, populating its relations.
4. Run the code. It should produce the error shown below.

**Expected behavior**
Populate should check if the connection / entity manager is currently inside a transaction, and if so it should not attempt to create its own and simply use the existing transactional context.

**Versions**

| Dependency | Version |
| - | - |
| node | 17/.1.0 |
| typescript | 4.3.5 |
| mikro-orm | 5.1.5 |
| @mikro-orm/postgresql | 5.1.5 |

**Comment by B4nan** [comment, marked as answer]
I cant help with such issues without a complete reproduction. You didnt even provided any code 🤷 

> Populate should check if the connection / entity manager is currently inside a transaction, and if so it should not attempt to create its own and simply use the existing transactional context.

Populate calls dont start any transactions. This error is not about starting a transaction, its about using a connection that is already closed.

> Additionally, the same error occurs when populate options are specified within a find() function or a findOne() function

Because that's exactly the same, both `find` and `findOne` are calling `em.populate` under the hood.

Share some code, ideally a complete one. And always upgrade to latest version before reporting anything (it probably won't help with this issue, but still...)

Maybe its the same problem as [this one](https://github.com/mikro-orm/mikro-orm/issues/3005)

**Reply by zdb999** [reply]
I spent some time today creating a repository to minimally recreate the issue, and was not able to do so successfully.

After further investigation, I found the source of my problem. I was not `await`ing one of my `ORM` calls inside the transaction, producing a race condition. Based on my loading strategy, `populate` caused an additional SQL query to be fired, resulting in the the `COMMIT` being reached first before the last SQL query. I believe this out of order execution produced the error shown above; adding the missing `await` fixed the issue.

I apologize for not providing a working code example the first time round, and for not vetting the issue further before posting.

Thank you for your time, and for making such an amazing piece of open source software.
